sys/net/gnrc/transport_layer/tcp/gnrc_tcp_fsm.c

```diff
@@ -0,0 +1,795 @@
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     net_gnrc
+ * @{
+ *
+ * @file
+ * @brief       GNRC's TCP finite state maschine
+ *
+ * @author      Simon Brummer <brummer.simon@googlemail.com>
+ * @}
+ */
+
+#include "msg.h"
+#include "random.h"
+#include "ringbuffer.h"
+#include "net/af.h"
+
+#ifdef MODULE_GNRC_IPV6
+#include "net/gnrc/ipv6.h"
+#endif
+
+#include "internal/fsm.h"
+#include "internal/pkt.h"
+#include "internal/option.h"
+#include "internal/helper.h"
+#include "internal/rcvbuf.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+/**
+ * @brief Checks if a given portnumber is currently used by a tcb as local_port.
+ *
+ * @param[in] portnumber   Portnumber that should be checked
+ *
+ * @note Must be called from a context where the tcb list ist locked.
+ *
+ * @return   Zero if @p portnumber is currently not used.
+ * @return   1 if @p portnumber is used by an tcb.
+ */
+static uint8_t _is_local_port_in_use(const uint16_t portnumber)
+{
+    gnrc_tcp_tcb_t *iter = NULL;
+    LL_FOREACH(_list_gnrc_tcp_tcb_head, iter) {
+        if (*(iter->local_port) == portnumber) {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+/**
+ * @brief Generate random, currently unused local port above the well-known ports (> 1024)
+ *
+ * @return The generated port number
+ */
+static uint16_t _get_random_local_port(void)
+{
+    uint16_t ret = 0;
+    do {
+        ret = random_uint32();
+        if (ret < 1024) {
+            continue;
+        }
+    } while(_is_local_port_in_use(ret));
+    return ret;
+}
+
+/**
+ * @brief clears retransmit queue
+ *
+ * @param[in/out] conn   TCP Connection, where the retransmit should be cleared
+ *
+ * @return zero on success
+ */
+static int8_t _clear_retransmit(gnrc_tcp_tcb_t *tcb)
+{
+    if (tcb->pkt_retransmit != NULL) {
+        gnrc_pktbuf_release(tcb->pkt_retransmit);
+        xtimer_remove(&(tcb->tim_tout));
+        tcb->pkt_retransmit = NULL;
+    }
+    return 0;
+}
+
+/**
+ * @brief restarts time wait timer
+ *
+ * @param[in/out] conn TCP Connection, where the timewait_timer should be restarted
+ *
+ * @return Zero on success
+ */
+static int8_t _restart_timewait_timer(gnrc_tcp_tcb_t* tcb)
+{
+    xtimer_remove(&tcb->tim_tout);
+    tcb->msg_tout.type = MSG_TYPE_TIMEWAIT;
+    tcb->msg_tout.content.ptr = (void *)tcb;
+    xtimer_set_msg(&tcb->tim_tout, 2 * GNRC_TCP_MSL, &tcb->msg_tout, _gnrc_tcp_pid);
+    return 0;
+}
+
+/**
+ * @brief translates fsm into another state
+ *
+ * @param[in/out] tcb            tcb, that specifies connection
+ * @param[in]     state          state to translate in
+ * @param[out]    notify_owner   non-negative if the tcb owner should be notified
+ *
+ * @return zero on success
+ */
+static int8_t _translate_to(gnrc_tcp_tcb_t* tcb, gnrc_tcp_fsm_state_t state, bool *notify_owner)
```

Comment:
Function name "_translate_to" changed to the mathematically correct term "_transition_to" in updated PR